### PR TITLE
Faster visit paging with bound and offset

### LINKED ISSUE #1019: Smarter pagination for history
Followup from #995, see https://github.com/mozilla/application-services/pull/995#issuecomment-483883406 for details. In particular, the current implementation is O(n^2) to iterate over all records.

We probably want to use some trickery to make things more efficient `OFFSET`, for example: https://github.com/mozilla-services/server-syncstorage/blob/9de1850199cff3a2771cbf752cf79a2391732456/syncstorage/storage/sql/__init__.py#L414

┆Issue is synchronized with this [Jira Story](https://jira.mozilla.com/browse/SYNC-77)

### PR DESCRIPTION
**Pull request by lyuyuan**
fixes #1019 
### Pull Request checklist ###

- [ ] **Quality**: This PR builds and tests run cleanly
  - `cargo test --all` produces no test failures
  - `cargo clippy --all --all-targets --all-features` runs without emitting any warnings
  - `cargo fmt` does not produce any changes to the code
  - `./gradlew ktlint detekt` runs without emitting any warnings
  - `swiftformat --swiftversion 4 megazords components/*/ios && swiftlint` runs without emitting any warnings or producing changes
  - Note: For changes that need extra cross-platform testing, consider adding `[ci full]` to the PR title.
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes a changelog entry in [CHANGES_UNRELEASED.md](../CHANGES_UNRELEASED.md) or an explanation of why it does not need one
  - Any breaking changes to Swift or Kotlin binding APIs are noted explicitly
- [ ] **Dependencies**: This PR follows our [dependency management guidelines](https://github.com/mozilla/application-services/blob/master/docs/dependency-management.md)
  - Any new dependencies are accompanied by a summary of the due dilligence applied in selecting them.

### CODE REVIEW

File: components/places/android/src/main/java/mozilla/appservices/places/LibPlacesFFI.kt [hunk 1]
```diff
@@ -161,6 +161,15 @@ internal interface LibPlacesFFI : Library {
         error: RustError.ByReference
     ): RustBuffer.ByValue
 
+    fun places_get_visit_page_with_bound(
+        handle: PlacesConnectionHandle,
+        offset: Long,
+        bound: Long,
```
Comment (rfk) [inline components/places/android/src/main/java/mozilla/appservices/places/LibPlacesFFI.kt]: nit: I think it would be better to switch the order of these parameters, e.g. pass in `bound, offset` rather than `offset, bound`. It will help the caller to form a more accurate mental model of how the underlying query works; "first skip directly to `bound`, then skip over `offset`  items from there".

File: components/places/android/src/main/java/mozilla/appservices/places/PlacesConnection.kt [hunk 2]
```diff
@@ -313,6 +313,24 @@ open class PlacesReaderConnection internal constructor(connHandle: Long) :
...
+            LibPlacesFFI.INSTANCE.places_get_visit_page_with_bound(
+                    this.handle.get(), offset, bound, count, visitTransitionSet(excludeTypes), error)
+        }
+        try {
+            val infos = MsgTypes.HistoryVisitInfos.parseFrom(infoBuffer.asCodedInputStream()!!)
+            return VisitInfo.fromMessage(infos)
+        } finally {
+            LibPlacesFFI.INSTANCE.places_destroy_bytebuffer(infoBuffer)
+        }
+    }
```
Comment (rfk) [inline components/places/android/src/main/java/mozilla/appservices/places/PlacesConnection.kt]: As a caller of this function, I would expect to be calling it in order to do some sort of traversal of multiple pages. How do I know what the right value of `bound` is to pass in?  For example, how do I write this loop:

```
bound = offset = 0
while (more pages) {
  page = getVisitPageWithBound(bound, offset, N, EXCLUDE)
  // do something with page
  bound = ??? // what's the next value of bound?
  offset = ???  // what's the next value of offset?
}
```

I expect that the caller will need to calculate them from the list of `VisitInfo`s that was returned, but we need to provide some instructions for how (or somehow return it from the previous call).

File: components/places/android/src/main/java/mozilla/appservices/places/PlacesConnection.kt [hunk 3]
```diff
@@ -704,6 +722,24 @@ interface ReadableHistoryConnection : InterruptibleConnection {
...
     fun getVisitPage(offset: Long, count: Long, excludeTypes: List<VisitType> = listOf()): List<VisitInfo>
 
+    /**
+     * Page more efficiently than using simple numeric offset. We first figure out
+     * a visited timestamp upper bound, then do a smaller numeric offset relative to
+     * the bound.
+     *
+     * @param offset The offset between first item that has visit date equal to bound
+     *  and last visited item.
+     * @param bound The upper bound of already visited items.
```
Comment (rfk) [inline components/places/android/src/main/java/mozilla/appservices/places/PlacesConnection.kt]: As above, I think switching the order of `offset` and `bound` will make it easier to form a correct mental model of the query here.

File: components/places/src/storage/history.rs [hunk 4]
```diff
@@ -1144,6 +1144,35 @@ pub fn get_visit_page(
...
+    exclude_types: VisitTransitionSet,
+) -> Result<HistoryVisitInfos> {
+    let allowed_types = exclude_types.complement();
+    let infos = db.query_rows_and_then_named_cached(
+        "SELECT h.url, h.title, v.visit_date, v.visit_type, h.hidden
+         FROM moz_places h
+         JOIN moz_historyvisits v
+           ON h.id = v.place_id
+         WHERE ((1 << v.visit_type) & :allowed_types) != 0 AND
+               NOT h.hidden AND v.visit_date <= :bound
```
Comment (rfk) [inline components/places/src/storage/history.rs]: nit: could you please put the `AND v.visit_date <= :bound` on its own line; it's pretty important to the query so making it more visually distinct seems helpful.

File: components/places/src/storage/history.rs [hunk 5]
```diff
@@ -1144,6 +1144,35 @@ pub fn get_visit_page(
...
+            ":allowed_types": allowed_types,
+            ":bound": bound,
+            ":count": count,
+            ":offset": offset,
+        },
+        HistoryVisitInfo::from_row,
+    )?;
+    Ok(HistoryVisitInfos { infos })
+}
+
```
Comment (rfk) [inline components/places/src/storage/history.rs]: Please also consider adding some tests for the new `get_visit_page_with_bound` function, copying the style of the tests further down in this file.  Interesting edge-cases to consider include:

* Does it work correctly when `count` is very small, and there are more than `count` items with the same `visit_date`?
* Does it work correctly when `offset` happens to fall precisely on a point where the `visited_date` changes? When it falls one item on either side?

File: components/places/android/src/main/java/mozilla/appservices/places/PlacesConnection.kt (hunk 2 above)
Comment (lyuyuan) [reply components/places/android/src/main/java/mozilla/appservices/places/PlacesConnection.kt]: Yes. I agree. And maybe return the next `bound` and `offset` can be even nicer than providing instructions?
So, if new bound is the same as previous bound, return the same `bound` with `offset` plus the size of page; If new bound is a different bound, return new `bound` with `offset=0`.
Does this sound right?

Comment (rfk) [reply components/places/android/src/main/java/mozilla/appservices/places/PlacesConnection.kt]: Yes, that sounds right. I agree that returning it somehow would be better, but it's not very obvious to me how to return it in a nice way over the FFI.

Comment (lyuyuan) [reply components/places/android/src/main/java/mozilla/appservices/places/PlacesConnection.kt]: I am thinking of creating a protobuf message type `OffsetToken` which contains the next `bound` and `offset` (both i64).

File: components/places/src/storage/history.rs (hunk 5 above)
Comment (lyuyuan) [reply components/places/src/storage/history.rs]: Create some tests under `test_get_visit_page_with_bound`.

File: components/places/src/storage/history.rs (hunk 4 above)
Comment (lyuyuan) [reply components/places/src/storage/history.rs]: Done.

File: components/places/android/src/main/java/mozilla/appservices/places/PlacesConnection.kt (hunk 3 above)
Comment (lyuyuan) [reply components/places/android/src/main/java/mozilla/appservices/places/PlacesConnection.kt]: Done.

File: components/places/android/src/main/java/mozilla/appservices/places/LibPlacesFFI.kt (hunk 1 above)
Comment (lyuyuan) [reply components/places/android/src/main/java/mozilla/appservices/places/LibPlacesFFI.kt]: Done.

File: CHANGES_UNRELEASED.md [hunk 6]
```diff
@@ -2,4 +2,12 @@
 
 # Unreleased Changes
 
+## Android
```
Comment (rfk) [inline CHANGES_UNRELEASED.md]: nit: I think the title here should be the name of the component, e.g. "Places" in this case. See also https://github.com/mozilla/application-services/pull/2314 which I just filed after noticing this here.

File: components/places/src/storage/history.rs [hunk 7]
```diff
@@ -1144,13 +1144,78 @@ pub fn get_visit_page(
...
+                infos,
+                bound,
+                offset,
+            })
+        } else {
+            let bound = l.timestamp;
+            let mut offset: i64 = 1;
+            let mut it = infos.iter().rev();
+            it.next();
+            for i in it {
```
Comment (rfk) [inline components/places/src/storage/history.rs]: nit: this loop can probably be replaced with a combination of the [take_while](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.take_while) and [count](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.count) iterator methods, which might turn out to be a bit more readable ("reverse the list, take elements while the timestamps equals the new bound, count them").

Just a suggestion though; doing it in a loop like this is also fine :-)

Comment (lyuyuan) [reply components/places/src/storage/history.rs]: Nice, it looks much nicer using `take_while` and `count`, thanks!

File: CHANGES_UNRELEASED.md (hunk 6 above)
Comment (lyuyuan) [reply CHANGES_UNRELEASED.md]: Corrected, thanks!
